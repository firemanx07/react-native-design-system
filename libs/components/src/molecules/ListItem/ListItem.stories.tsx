import { ComponentStory, ComponentMeta } from '@storybook/react';
import React from 'react';

import { BellIcon } from '../../atoms';
import { StorybookScreen, StorybookSection } from '../../storybook';
import {
  ListItem as ListItemBase,
  PropsType as ListItemProps,
} from './ListItem';

const Story: ComponentMeta<typeof ListItemBase> = {
  component: ListItemBase,
  title: 'Molecules/List Item',
};
export default Story;

const renderIconLeft = (size: number, color: string) => (
  <BellIcon width={size} height={size} fill={color} />
);

const listItemText = 'notifications';

const Template: ComponentStory<typeof ListItemBase> = (args: ListItemProps) => (
  <StorybookScreen title="ListItem:">
    <StorybookSection>
      <ListItemBase {...args} />
    </StorybookSection>
    <StorybookSection>
      <ListItemBase renderIconLeft={renderIconLeft} hasArrow>
        {listItemText}
      </ListItemBase>
    </StorybookSection>
    <StorybookSection>
      <ListItemBase renderIconLeft={renderIconLeft} hasArrow notifications={1}>
        {listItemText}
      </ListItemBase>
    </StorybookSection>
    <StorybookSection>
      <ListItemBase renderIconLeft={renderIconLeft}>
        {listItemText}
      </ListItemBase>
    </StorybookSection>
    <StorybookSection>
      <ListItemBase renderIconLeft={renderIconLeft} notifications={1}>
        {listItemText}
      </ListItemBase>
    </StorybookSection>
  </StorybookScreen>
);

const parameters = {
  controls: {
    include: ['hasArrow', 'notifications', 'children'],
  },
};

const defaultArgs = {
  renderIconLeft,
  hasArrow: true,
  notifications: 1,
  children: listItemText,
};

export const ListItem = Template.bind({});
ListItem.parameters = parameters;
ListItem.args = {
  ...defaultArgs,
};
