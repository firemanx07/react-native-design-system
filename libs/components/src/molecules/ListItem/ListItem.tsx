import { styled, useNamespacedTheme } from '@proxym/themes';
import React, { memo, ReactNode } from 'react';
import { ViewStyle, StyleProp, TextProps } from 'react-native';

import {
  Badge,
  BadgeVariant,
  BaseText,
  ForwardIcon,
  TextVariant,
  TextWeight,
} from '../../atoms';
import { TestIDType } from '../../types';

export type PropsType = {
  children: TextProps['children'];
  style?: StyleProp<ViewStyle>;
  renderIconLeft?: (size: number, color: string) => ReactNode;
  hasArrow?: boolean;
  notifications?: number;
} & TestIDType;

export const ListItem = ({
  style,
  renderIconLeft,
  hasArrow,
  notifications,
  children,
  testID,
}: PropsType) => {
  const { colors, iconSize } = useNamespacedTheme();

  return (
    <Container style={style} testID={testID}>
      {!!renderIconLeft && (
        <IconLeftWrapper>
          {renderIconLeft(iconSize.primary, colors.dark)}
        </IconLeftWrapper>
      )}
      <TextWrapper>
        <BaseText
          variant={TextVariant.L}
          weight={TextWeight.SemiBold}
          numberOfLines={2}
        >
          {children}
        </BaseText>
      </TextWrapper>
      {!!notifications && (
        <NotificationsBadge variant={BadgeVariant.Rounded} hasArrow={hasArrow}>
          {notifications}
        </NotificationsBadge>
      )}
      {hasArrow && (
        <ForwardIcon
          width={iconSize.medium}
          height={iconSize.medium}
          fill={colors.dark}
        />
      )}
    </Container>
  );
};

const Container = styled.View`
  flex-direction: row;
  align-items: center;
  background-color: ${({ theme }) => theme.ds.colors.light};
  padding-horizontal: ${({ theme }) => theme.ds.spacing.primary}px;
  min-height: 48px;
`;

const IconLeftWrapper = styled.View`
  padding-left: ${({ theme }) => theme.ds.spacing.tiny}px;
  padding-right: ${({ theme }) => theme.ds.spacing.primarySmall}px;
`;

const TextWrapper = styled.View`
  flex: 1;
`;

type NotificationBadgeType = {
  hasArrow?: boolean;
};

const NotificationsBadge = styled(Badge)<NotificationBadgeType>`
  margin-left: ${({ theme }) => theme.ds.spacing.primary}px;
  margin-right: ${({ theme, hasArrow }) =>
    hasArrow ? theme.ds.spacing.primary : theme.ds.spacing.small}px;
  align-self: center;
`;

export default memo(ListItem);
