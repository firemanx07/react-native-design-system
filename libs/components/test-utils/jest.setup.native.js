import { Animated } from 'react-native';

require('react-native-reanimated/lib/reanimated2/jestUtils').setUpTests();

console.warn = () => {};

Animated.timing = () => ({
  start: () => jest.fn(),
});
