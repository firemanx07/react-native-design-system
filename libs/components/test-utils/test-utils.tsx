import { theme, ThemeProvider } from '@celebrate-app/themes';
import { Story } from '@storybook/react';
import {
  cleanup,
  render,
  RenderOptions,
  waitFor,
} from '@testing-library/react-native';
import React, { ReactNode } from 'react';

const Providers = ({ children }: { children: ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export const customRender = (
  ui: React.ReactElement<any>,
  options?: RenderOptions
) => {
  return render(ui, { wrapper: Providers, ...options });
};

beforeEach(cleanup);
afterEach(cleanup);

export function testStory(displayName: string, TestComponent: Story) {
  test(`${displayName} snapshot test`, async () => {
    const comp = customRender(<TestComponent {...TestComponent.args} />);
    jest.runOnlyPendingTimers();

    await waitFor(() => {
      expect(comp).toBeTruthy();
      expect(comp.toJSON()).toMatchSnapshot();
    });
  });
}

export function testComponentSnapshot(
  TestComponent: React.ReactElement<
    any,
    string | React.JSXElementConstructor<any>
  >
) {
  test('Snapshot test', () => {
    const comp = customRender(TestComponent);
    expect(comp).toBeTruthy();
    expect(comp.toJSON()).toMatchSnapshot();
  });
}

export function testAllStories(stories: any) {
  Object.keys(stories).forEach((story) => {
    if (story !== 'default') {
      testStory(story, stories[story]);
    }
  });
}
