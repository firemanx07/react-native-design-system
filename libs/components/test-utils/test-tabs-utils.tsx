import { waitFor } from '@testing-library/react-native';

import { customRender } from './test-utils';

type renderTabsProps = {
  elem: JSX.Element;
  testIDs: string[];
};
export const renderTabs = ({ elem, testIDs }: renderTabsProps) => {
  const comp = customRender(elem);
  const tabs: ReturnType<typeof comp.getByTestId>[] = [];
  testIDs.forEach(testID => tabs.push(comp.getByTestId(testID)));
  return { comp, tabs };
};
export const testTabs = async (args: ReturnType<typeof renderTabs>) => {
  await waitFor(() => {
    expect(args.comp).toBeTruthy();
    args.tabs.forEach(tab => expect(tab).toBeDefined());
  });
};
