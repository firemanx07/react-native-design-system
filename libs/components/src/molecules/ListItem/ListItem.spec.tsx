import { testAllStories } from '../../../test-utils';
import * as stories from './ListItem.stories';

describe('ListItem tests', () => {
  testAllStories(stories);
});
