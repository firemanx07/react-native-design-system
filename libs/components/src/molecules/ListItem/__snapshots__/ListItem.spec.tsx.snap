// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem tests ListItem snapshot test 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F2F3F4",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "minHeight": 600,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      },
    ]
  }
>
  <Text
    fontFamily="Muli-Bold"
    fontSize={20}
    lineHeight={26}
    style={
      [
        {
          "color": "#222222",
          "fontFamily": "Muli-Bold",
          "fontSize": 20,
          "lineHeight": 26,
        },
        {
          "marginBottom": 24,
        },
      ]
    }
  >
    ListItem:
  </Text>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "flexDirection": "row",
            "minHeight": 48,
            "paddingHorizontal": 16,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "paddingLeft": 4,
              "paddingRight": 12,
            },
          ]
        }
      >
        <test-file-stub
          fill="#222222"
          height={24}
          width={24}
        />
      </View>
      <View
        style={
          [
            {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      >
        <Text
          fontFamily="Muli-SemiBold"
          fontSize={16}
          lineHeight={20}
          numberOfLines={2}
          style={
            [
              {
                "color": "#222222",
                "fontFamily": "Muli-SemiBold",
                "fontSize": 16,
                "lineHeight": 20,
              },
            ]
          }
        >
          Benachrichtigungen
        </Text>
      </View>
      <View
        borderRadius={12}
        size={24}
        style={
          [
            {
              "alignItems": "center",
              "alignSelf": "flex-start",
              "backgroundColor": "#FF8066",
              "borderRadius": 12,
              "flexDirection": "row",
              "height": 24,
              "justifyContent": "center",
              "minWidth": 24,
              "paddingBottom": 4,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 4,
            },
            {
              "alignSelf": "center",
              "marginLeft": 16,
              "marginRight": 16,
            },
          ]
        }
      >
        <Text
          color="#FFFFFF"
          fontFamily="Muli-Bold"
          fontSize={14}
          lineHeight={20}
          style={
            [
              {
                "color": "#222222",
                "fontFamily": "Muli-Bold",
                "fontSize": 14,
                "lineHeight": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        >
          1
        </Text>
      </View>
      <test-file-stub
        fill="#222222"
        height={20}
        width={20}
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "flexDirection": "row",
            "minHeight": 48,
            "paddingHorizontal": 16,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "paddingLeft": 4,
              "paddingRight": 12,
            },
          ]
        }
      >
        <test-file-stub
          fill="#222222"
          height={24}
          width={24}
        />
      </View>
      <View
        style={
          [
            {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      >
        <Text
          fontFamily="Muli-SemiBold"
          fontSize={16}
          lineHeight={20}
          numberOfLines={2}
          style={
            [
              {
                "color": "#222222",
                "fontFamily": "Muli-SemiBold",
                "fontSize": 16,
                "lineHeight": 20,
              },
            ]
          }
        >
          Benachrichtigungen
        </Text>
      </View>
      <test-file-stub
        fill="#222222"
        height={20}
        width={20}
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "flexDirection": "row",
            "minHeight": 48,
            "paddingHorizontal": 16,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "paddingLeft": 4,
              "paddingRight": 12,
            },
          ]
        }
      >
        <test-file-stub
          fill="#222222"
          height={24}
          width={24}
        />
      </View>
      <View
        style={
          [
            {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      >
        <Text
          fontFamily="Muli-SemiBold"
          fontSize={16}
          lineHeight={20}
          numberOfLines={2}
          style={
            [
              {
                "color": "#222222",
                "fontFamily": "Muli-SemiBold",
                "fontSize": 16,
                "lineHeight": 20,
              },
            ]
          }
        >
          Benachrichtigungen
        </Text>
      </View>
      <View
        borderRadius={12}
        size={24}
        style={
          [
            {
              "alignItems": "center",
              "alignSelf": "flex-start",
              "backgroundColor": "#FF8066",
              "borderRadius": 12,
              "flexDirection": "row",
              "height": 24,
              "justifyContent": "center",
              "minWidth": 24,
              "paddingBottom": 4,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 4,
            },
            {
              "alignSelf": "center",
              "marginLeft": 16,
              "marginRight": 16,
            },
          ]
        }
      >
        <Text
          color="#FFFFFF"
          fontFamily="Muli-Bold"
          fontSize={14}
          lineHeight={20}
          style={
            [
              {
                "color": "#222222",
                "fontFamily": "Muli-Bold",
                "fontSize": 14,
                "lineHeight": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        >
          1
        </Text>
      </View>
      <test-file-stub
        fill="#222222"
        height={20}
        width={20}
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "flexDirection": "row",
            "minHeight": 48,
            "paddingHorizontal": 16,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "paddingLeft": 4,
              "paddingRight": 12,
            },
          ]
        }
      >
        <test-file-stub
          fill="#222222"
          height={24}
          width={24}
        />
      </View>
      <View
        style={
          [
            {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      >
        <Text
          fontFamily="Muli-SemiBold"
          fontSize={16}
          lineHeight={20}
          numberOfLines={2}
          style={
            [
              {
                "color": "#222222",
                "fontFamily": "Muli-SemiBold",
                "fontSize": 16,
                "lineHeight": 20,
              },
            ]
          }
        >
          Benachrichtigungen
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      [
        {
          "borderColor": "#222222",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "flexDirection": "row",
            "minHeight": 48,
            "paddingHorizontal": 16,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "paddingLeft": 4,
              "paddingRight": 12,
            },
          ]
        }
      >
        <test-file-stub
          fill="#222222"
          height={24}
          width={24}
        />
      </View>
      <View
        style={
          [
            {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      >
        <Text
          fontFamily="Muli-SemiBold"
          fontSize={16}
          lineHeight={20}
          numberOfLines={2}
          style={
            [
              {
                "color": "#222222",
                "fontFamily": "Muli-SemiBold",
                "fontSize": 16,
                "lineHeight": 20,
              },
            ]
          }
        >
          Benachrichtigungen
        </Text>
      </View>
      <View
        borderRadius={12}
        size={24}
        style={
          [
            {
              "alignItems": "center",
              "alignSelf": "flex-start",
              "backgroundColor": "#FF8066",
              "borderRadius": 12,
              "flexDirection": "row",
              "height": 24,
              "justifyContent": "center",
              "minWidth": 24,
              "paddingBottom": 4,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 4,
            },
            {
              "alignSelf": "center",
              "marginLeft": 16,
              "marginRight": 8,
            },
          ]
        }
      >
        <Text
          color="#FFFFFF"
          fontFamily="Muli-Bold"
          fontSize={14}
          lineHeight={20}
          style={
            [
              {
                "color": "#222222",
                "fontFamily": "Muli-Bold",
                "fontSize": 14,
                "lineHeight": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        >
          1
        </Text>
      </View>
    </View>
  </View>
</View>
`;
